File: tom_process_monitor/lib/src/client/process_monitor_client.dart
import 'dart:io';

import 'package:path/path.dart' as path;

import '../exceptions/process_disabled_exception.dart';
import '../exceptions/process_monitor_exception.dart';
import '../exceptions/process_not_found_exception.dart';
import '../models/partner_discovery_config.dart';
import '../models/process_config.dart';
import '../models/process_entry.dart';
import '../models/process_state.dart';
import '../models/process_status.dart';
import '../models/remote_access_config.dart';
import '../services/process_control.dart';
import '../services/registry_service.dart';

/// Local client API for interacting with ProcessMonitor.
///
/// This client communicates via the file-based registry and does not
/// require direct connection to the ProcessMonitor daemon.
class ProcessMonitorClient {
  /// Directory containing registry and lock files.
  final String directory;

  /// ProcessMonitor instance ID.
  final String instanceId;

  late final RegistryService _registry;
  late final ProcessControl _processControl;

  /// Creates a local process monitor client.
  ProcessMonitorClient({
    String? directory,
    this.instanceId = 'default',
  }) : directory = directory ?? _resolveDefaultDirectory() {
    _registry = RegistryService(
      directory: this.directory,
      instanceId: instanceId,
    );
    _processControl = ProcessControl(
      logDirectory: this.directory,
    );
  }

  // --- Registration ---

  /// Register a new local process with the monitor.
  Future<void> register(ProcessConfig config) async {
    await _registry.withLock((registry) async {
      if (registry.processes.containsKey(config.id)) {
        throw ProcessMonitorException('Process ${config.id} already exists');
      }

      registry.processes[config.id] = ProcessEntry(
        id: config.id,
        name: config.name,
        command: config.command,
        args: config.args,
        workingDirectory: config.workingDirectory,
        environment: config.environment,
        autostart: config.autostart,
        enabled: true,
        isRemote: false,
        restartPolicy: config.restartPolicy,
        alivenessCheck: config.alivenessCheck,
        registeredAt: DateTime.now(),
      );
    });
  }

  /// Remove a process from the registry.
  /// Stops the process if running.
  Future<void> deregister(String processId) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      // Stop if running
      if (process.pid != null && process.state == ProcessState.running) {
        await _processControl.stopProcessGracefully(process.pid!);
      }

      registry.processes.remove(processId);
    });
  }

  // --- Enable/Disable ---

  /// Enable a process (allows it to be started).
  Future<void> enable(String processId) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      process.enabled = true;
      if (process.state == ProcessState.disabled) {
        process.state = ProcessState.stopped;
      }
    });
  }

  /// Disable a process (stops it and prevents restart).
  Future<void> disable(String processId) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      // Stop if running
      if (process.pid != null && process.state == ProcessState.running) {
        await _processControl.stopProcessGracefully(process.pid!);
        process.pid = null;
        process.lastStoppedAt = DateTime.now();
      }

      process.enabled = false;
      process.state = ProcessState.disabled;
    });
  }

  // --- Autostart ---

  /// Set whether the process starts automatically.
  Future<void> setAutostart(String processId, bool autostart) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      process.autostart = autostart;
    });
  }

  // --- Process Control ---

  /// Start a process (if enabled).
  Future<void> start(String processId) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      if (!process.enabled) {
        throw ProcessDisabledException(processId);
      }

      if (process.state == ProcessState.running && process.pid != null) {
        // Already running
        return;
      }

      process.state = ProcessState.starting;
    });
  }

  /// Stop a process (does not disable it).
  Future<void> stop(String processId) async {
    await _registry.withLock((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      if (process.pid != null) {
        process.state = ProcessState.stopping;
        await _processControl.stopProcessGracefully(process.pid!);
        process.pid = null;
        process.lastStoppedAt = DateTime.now();
      }

      process.state = ProcessState.stopped;
    });
  }

  /// Restart a process (stop then start).
  Future<void> restart(String processId) async {
    await stop(processId);
    await start(processId);
  }

  // --- Status ---

  /// Get status of a specific process.
  Future<ProcessStatus> getStatus(String processId) async {
    return _registry.withLockReadOnly((registry) async {
      final process = registry.processes[processId];
      if (process == null) {
        throw ProcessNotFoundException(processId);
      }

      return _toStatus(process);
    });
  }

  /// Get status of all registered processes.
  Future<Map<String, ProcessStatus>> getAllStatus() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.processes.map(
        (key, value) => MapEntry(key, _toStatus(value)),
      );
    });
  }

  // --- Remote Access Configuration ---

  /// Enable or disable remote HTTP API access.
  Future<void> setRemoteAccess(bool enabled) async {
    await _registry.withLock((registry) async {
      registry.remoteAccess = registry.remoteAccess.copyWith(
        startRemoteAccess: enabled,
      );
    });
  }

  /// Get current remote access configuration.
  Future<RemoteAccessConfig> getRemoteAccessConfig() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.remoteAccess;
    });
  }

  /// Set remote access permissions.
  Future<void> setRemoteAccessPermissions({
    bool? allowRegister,
    bool? allowDeregister,
    bool? allowStart,
    bool? allowStop,
    bool? allowDisable,
    bool? allowAutostart,
    bool? allowMonitorRestart,
  }) async {
    await _registry.withLock((registry) async {
      registry.remoteAccess = registry.remoteAccess.copyWith(
        allowRemoteRegister: allowRegister,
        allowRemoteDeregister: allowDeregister,
        allowRemoteStart: allowStart,
        allowRemoteStop: allowStop,
        allowRemoteDisable: allowDisable,
        allowRemoteAutostart: allowAutostart,
        allowRemoteMonitorRestart: allowMonitorRestart,
      );
    });
  }

  /// Set trusted hosts list.
  Future<void> setTrustedHosts(List<String> hosts) async {
    await _registry.withLock((registry) async {
      registry.remoteAccess = registry.remoteAccess.copyWith(
        trustedHosts: hosts,
      );
    });
  }

  /// Get trusted hosts list.
  Future<List<String>> getTrustedHosts() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.remoteAccess.trustedHosts;
    });
  }

  // --- Executable Filtering ---

  /// Get the current executable whitelist.
  Future<List<String>> getRemoteExecutableWhitelist() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.remoteAccess.executableWhitelist;
    });
  }

  /// Set the executable whitelist (glob patterns).
  Future<void> setRemoteExecutableWhitelist(List<String> patterns) async {
    await _registry.withLock((registry) async {
      registry.remoteAccess = registry.remoteAccess.copyWith(
        executableWhitelist: patterns,
      );
    });
  }

  /// Get the current executable blacklist.
  Future<List<String>> getRemoteExecutableBlacklist() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.remoteAccess.executableBlacklist;
    });
  }

  /// Set the executable blacklist (glob patterns).
  Future<void> setRemoteExecutableBlacklist(List<String> patterns) async {
    await _registry.withLock((registry) async {
      registry.remoteAccess = registry.remoteAccess.copyWith(
        executableBlacklist: patterns,
      );
    });
  }

  // --- Standalone / Partner Configuration ---

  /// Enable or disable standalone mode (no partner monitoring).
  Future<void> setStandaloneMode(bool enabled) async {
    await _registry.withLock((registry) async {
      registry.standaloneMode = enabled;
    });
  }

  /// Get current standalone mode setting.
  Future<bool> isStandaloneMode() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.standaloneMode;
    });
  }

  /// Get partner discovery configuration.
  Future<PartnerDiscoveryConfig> getPartnerDiscoveryConfig() async {
    return _registry.withLockReadOnly((registry) async {
      return registry.partnerDiscovery;
    });
  }

  /// Set partner discovery configuration.
  Future<void> setPartnerDiscoveryConfig(PartnerDiscoveryConfig config) async {
    await _registry.withLock((registry) async {
      registry.partnerDiscovery = config;
    });
  }

  // --- Monitor Control ---

  /// Restart the ProcessMonitor itself.
  /// This sets a flag that the monitor will pick up.
  Future<void> restartMonitor() async {
    // Create a restart signal file
    final signalFile = File(path.join(directory, 'restart_$instanceId.signal'));
    await signalFile.writeAsString(DateTime.now().toIso8601String());
  }

  ProcessStatus _toStatus(ProcessEntry entry) {
    return ProcessStatus(
      id: entry.id,
      name: entry.name,
      state: entry.state,
      enabled: entry.enabled,
      autostart: entry.autostart,
      isRemote: entry.isRemote,
      pid: entry.pid,
      lastStartedAt: entry.lastStartedAt,
      lastStoppedAt: entry.lastStoppedAt,
      restartAttempts: entry.restartAttempts,
    );
  }
}

/// Resolves the user home directory cross-platform.
String _resolveHomeDirectory() {
  // Try HOME first (Linux, macOS)
  final home = Platform.environment['HOME'];
  if (home != null) return home;

  // Try USERPROFILE (Windows)
  final userProfile = Platform.environment['USERPROFILE'];
  if (userProfile != null) return userProfile;

  // Fallback to current directory
  return '.';
}

/// Resolves the default directory based on context.
String _resolveDefaultDirectory() {
  // In VS Code context: workspace root
  // Outside VS Code: user home directory
  final vsCodeWorkspace = Platform.environment['VSCODE_WORKSPACE_FOLDER'];
  if (vsCodeWorkspace != null) {
    return path.join(vsCodeWorkspace, '.tom', 'process_monitor');
  }
  return path.join(
    _resolveHomeDirectory(),
    '.tom',
    'process_monitor',
  );
}

File: tom_distributed_typescript/src/process_monitor_client.ts
/**
 * TypeScript client for ProcessMonitor.
 * 
 * Provides remote access to ProcessMonitor via HTTP API.
 * Mirrors the Dart RemoteProcessMonitorClient implementation.
 */

import { HttpClient, HttpClientOptions, checkResponse, HttpException, RetryExhaustedException } from './http_client';
import { discover, DiscoveryFailedException, scanSubnet } from './discovery';

export { DiscoveryFailedException, RetryExhaustedException };

// ─────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────

/**
 * Process state enumeration.
 */
export type ProcessState = 
  | 'stopped'
  | 'running'
  | 'starting'
  | 'stopping'
  | 'crashed'
  | 'disabled'
  | 'unknown';

/**
 * Restart policy types.
 */
export type RestartPolicyType = 'always' | 'on-failure' | 'never';

/**
 * Restart policy configuration.
 */
export interface RestartPolicy {
  type: RestartPolicyType;
  maxAttempts?: number;
  delayMs?: number;
  maxDelayMs?: number;
  backoffMultiplier?: number;
}

/**
 * HTTP aliveness check configuration.
 */
export interface AlivenessCheck {
  url: string;
  intervalMs?: number;
  timeoutMs?: number;
  successCodes?: number[];
}

/**
 * Configuration for registering a process.
 */
export interface ProcessConfig {
  /** Unique identifier for the process. */
  id: string;
  
  /** Human-readable name. */
  name: string;
  
  /** Executable command. */
  command: string;
  
  /** Command-line arguments. */
  args?: string[];
  
  /** Working directory. */
  workingDirectory?: string;
  
  /** Environment variables. */
  environment?: Record<string, string>;
  
  /** Start automatically when ProcessMonitor initializes. */
  autostart?: boolean;
  
  /** Restart policy configuration. */
  restartPolicy?: RestartPolicy;
  
  /** HTTP aliveness check configuration. */
  alivenessCheck?: AlivenessCheck;
}

/**
 * Status information for a process.
 */
export interface ProcessStatus {
  /** Unique process identifier. */
  id: string;
  
  /** Human-readable process name. */
  name: string;
  
  /** Current process state. */
  state: ProcessState;
  
  /** Whether the process can be started. */
  enabled: boolean;
  
  /** Start on ProcessMonitor initialization. */
  autostart: boolean;
  
  /** Whether registered via remote API. */
  isRemote: boolean;
  
  /** Current process ID (if running). */
  pid?: number;
  
  /** When process was last started. */
  lastStartedAt?: Date;
  
  /** When process was last stopped. */
  lastStoppedAt?: Date;
  
  /** Current restart attempt count. */
  restartAttempts: number;
}

/**
 * Status information for the ProcessMonitor instance.
 */
export interface MonitorStatus {
  /** ProcessMonitor instance ID. */
  instanceId: string;
  
  /** Process ID of this instance. */
  pid: number;
  
  /** When this instance started. */
  startedAt: Date;
  
  /** Seconds since startup. */
  uptime: number;
  
  /** Current state ("running", "stopping"). */
  state: string;
  
  /** Whether running in standalone mode (no partner). */
  standaloneMode: boolean;
  
  /** Partner instance ID (null if standalone). */
  partnerInstanceId?: string;
  
  /** Partner status ("running", "stopped", "unknown"). */
  partnerStatus?: string;
  
  /** Partner's PID (null if unknown or standalone). */
  partnerPid?: number;
  
  /** Total number of managed processes. */
  managedProcessCount: number;
  
  /** Number of currently running processes. */
  runningProcessCount: number;
}

/**
 * Remote access configuration.
 */
export interface RemoteAccessConfig {
  startRemoteAccess: boolean;
  allowRemoteRegister: boolean;
  allowRemoteDeregister: boolean;
  allowRemoteStart: boolean;
  allowRemoteStop: boolean;
  allowRemoteDisable: boolean;
  allowRemoteAutostart: boolean;
  allowRemoteMonitorRestart: boolean;
}

/**
 * Partner discovery configuration.
 */
export interface PartnerDiscoveryConfig {
  enabled: boolean;
  scanIntervalMs?: number;
  subnets?: string[];
}

// ─────────────────────────────────────────────────────────────
// Exceptions
// ─────────────────────────────────────────────────────────────

/**
 * Exception thrown when a process is not found.
 */
export class ProcessNotFoundException extends Error {
  readonly processId: string;
  
  constructor(processId: string) {
    super(`Process not found: ${processId}`);
    this.name = 'ProcessNotFoundException';
    this.processId = processId;
  }
}

/**
 * Exception thrown when permission is denied.
 */
export class PermissionDeniedException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PermissionDeniedException';
  }
}

// ─────────────────────────────────────────────────────────────
// Client Options
// ─────────────────────────────────────────────────────────────

/**
 * Options for creating a ProcessMonitorClient.
 */
export interface ProcessMonitorClientOptions extends HttpClientOptions {
  /** Target ProcessMonitor instance ID. Default: 'default' */
  instanceId?: string;
}

// ─────────────────────────────────────────────────────────────
// Client Implementation
// ─────────────────────────────────────────────────────────────

/** Default ProcessMonitor port. */
export const DEFAULT_PROCESS_MONITOR_PORT = 19881;

/**
 * Remote client for ProcessMonitor.
 * 
 * Provides HTTP-based access to all ProcessMonitor operations.
 */
export class ProcessMonitorClient {
  private readonly client: HttpClient;
  
  /** The base URL of the ProcessMonitor. */
  readonly baseUrl: string;
  
  /** The target ProcessMonitor instance ID. */
  readonly instanceId: string;
  
  /**
   * Creates a new ProcessMonitorClient.
   * 
   * @param options Client options
   */
  constructor(options: ProcessMonitorClientOptions = {}) {
    this.baseUrl = options.baseUrl ?? `http://localhost:${DEFAULT_PROCESS_MONITOR_PORT}`;
    this.instanceId = options.instanceId ?? 'default';
    this.client = new HttpClient({
      baseUrl: this.baseUrl,
      enableRetry: options.enableRetry ?? true,
      retryConfig: options.retryConfig,
      timeout: options.timeout,
    });
  }
  
  /**
   * Auto-discover a ProcessMonitor instance.
   * 
   * Discovery order:
   * 1. Try localhost on default port (19881)
   * 2. Try 127.0.0.1 on default port
   * 3. Try all local machine IP addresses
   * 4. Scan all /24 subnets for each local IP
   * 
   * @param options Discovery options
   */
  static async discover(options: {
    port?: number;
    timeout?: number;
    instanceId?: string;
  } = {}): Promise<ProcessMonitorClient> {
    const port = options.port ?? DEFAULT_PROCESS_MONITOR_PORT;
    
    const result = await discover({
      port,
      timeout: options.timeout ?? 5000,
      healthPath: '/monitor/status',
    });
    
    return new ProcessMonitorClient({
      baseUrl: result.url,
      instanceId: options.instanceId ?? 'default',
    });
  }
  
  /**
   * Scan a subnet for ProcessMonitor instances.
   * 
   * @param subnet First 3 octets (e.g., "192.168.1")
   * @param options Scan options
   */
  static async scanSubnet(
    subnet: string,
    options: { port?: number; timeout?: number } = {},
  ): Promise<string[]> {
    const port = options.port ?? DEFAULT_PROCESS_MONITOR_PORT;
    return scanSubnet(subnet, port, '/monitor/status', options.timeout ?? 500);
  }
  
  /**
   * Disposes the client.
   */
  dispose(): void {
    this.client.dispose();
  }
  
  // ─────────────────────────────────────────────────────────────
  // Registration
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Register a new process.
   */
  async register(config: ProcessConfig): Promise<void> {
    const response = await this.client.post('/processes', config);
    this.checkResponse(response);
  }
  
  /**
   * Remove a process from the registry.
   */
  async deregister(processId: string): Promise<void> {
    const response = await this.client.delete(`/processes/${processId}`);
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Enable/Disable
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Enable a process.
   */
  async enable(processId: string): Promise<void> {
    const response = await this.client.post(`/processes/${processId}/enable`);
    this.checkResponse(response);
  }
  
  /**
   * Disable a process.
   */
  async disable(processId: string): Promise<void> {
    const response = await this.client.post(`/processes/${processId}/disable`);
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Autostart
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Set autostart for a process.
   */
  async setAutostart(processId: string, autostart: boolean): Promise<void> {
    const response = await this.client.put(
      `/processes/${processId}/autostart`,
      { autostart },
    );
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Process Control
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Start a process.
   */
  async start(processId: string): Promise<void> {
    const response = await this.client.post(`/processes/${processId}/start`);
    this.checkResponse(response);
  }
  
  /**
   * Stop a process.
   */
  async stop(processId: string): Promise<void> {
    const response = await this.client.post(`/processes/${processId}/stop`);
    this.checkResponse(response);
  }
  
  /**
   * Restart a process.
   */
  async restart(processId: string): Promise<void> {
    const response = await this.client.post(`/processes/${processId}/restart`);
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Status
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Get status of a specific process.
   */
  async getStatus(processId: string): Promise<ProcessStatus> {
    const response = await this.client.get(`/processes/${processId}`);
    this.checkResponse(response);
    return this.parseProcessStatus(JSON.parse(response.body));
  }
  
  /**
   * Get status of all processes.
   */
  async getAllStatus(): Promise<Map<string, ProcessStatus>> {
    const response = await this.client.get('/processes');
    this.checkResponse(response);
    
    const data = JSON.parse(response.body);
    const processes = data.processes as unknown[];
    const result = new Map<string, ProcessStatus>();
    
    for (const p of processes) {
      const status = this.parseProcessStatus(p as Record<string, unknown>);
      result.set(status.id, status);
    }
    
    return result;
  }
  
  /**
   * Get ProcessMonitor instance status.
   */
  async getMonitorStatus(): Promise<MonitorStatus> {
    const response = await this.client.get('/monitor/status');
    this.checkResponse(response);
    return this.parseMonitorStatus(JSON.parse(response.body));
  }
  
  // ─────────────────────────────────────────────────────────────
  // Remote Access Configuration
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Enable or disable remote HTTP API access.
   */
  async setRemoteAccess(enabled: boolean): Promise<void> {
    const response = await this.client.put('/config/remote-access', {
      startRemoteAccess: enabled,
    });
    this.checkResponse(response);
  }
  
  /**
   * Get current remote access configuration.
   */
  async getRemoteAccessConfig(): Promise<RemoteAccessConfig> {
    const response = await this.client.get('/config/remote-access');
    this.checkResponse(response);
    return JSON.parse(response.body) as RemoteAccessConfig;
  }
  
  /**
   * Set remote access permissions.
   */
  async setRemoteAccessPermissions(permissions: Partial<RemoteAccessConfig>): Promise<void> {
    const response = await this.client.put('/config/remote-access', permissions);
    this.checkResponse(response);
  }
  
  /**
   * Set trusted hosts list.
   */
  async setTrustedHosts(hosts: string[]): Promise<void> {
    const response = await this.client.put('/config/trusted-hosts', {
      trustedHosts: hosts,
    });
    this.checkResponse(response);
  }
  
  /**
   * Get trusted hosts list.
   */
  async getTrustedHosts(): Promise<string[]> {
    const response = await this.client.get('/config/trusted-hosts');
    this.checkResponse(response);
    const data = JSON.parse(response.body);
    return data.trustedHosts as string[];
  }
  
  // ─────────────────────────────────────────────────────────────
  // Executable Filtering
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Get the current executable whitelist.
   */
  async getExecutableWhitelist(): Promise<string[]> {
    const response = await this.client.get('/config/executable-whitelist');
    this.checkResponse(response);
    const data = JSON.parse(response.body);
    return data.patterns as string[];
  }
  
  /**
   * Set the executable whitelist.
   */
  async setExecutableWhitelist(patterns: string[]): Promise<void> {
    const response = await this.client.put('/config/executable-whitelist', {
      patterns,
    });
    this.checkResponse(response);
  }
  
  /**
   * Get the current executable blacklist.
   */
  async getExecutableBlacklist(): Promise<string[]> {
    const response = await this.client.get('/config/executable-blacklist');
    this.checkResponse(response);
    const data = JSON.parse(response.body);
    return data.patterns as string[];
  }
  
  /**
   * Set the executable blacklist.
   */
  async setExecutableBlacklist(patterns: string[]): Promise<void> {
    const response = await this.client.put('/config/executable-blacklist', {
      patterns,
    });
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Standalone / Partner Configuration
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Enable or disable standalone mode.
   */
  async setStandaloneMode(enabled: boolean): Promise<void> {
    const response = await this.client.put('/config/standalone-mode', {
      enabled,
    });
    this.checkResponse(response);
  }
  
  /**
   * Get current standalone mode setting.
   */
  async isStandaloneMode(): Promise<boolean> {
    const response = await this.client.get('/config/standalone-mode');
    this.checkResponse(response);
    const data = JSON.parse(response.body);
    return data.enabled as boolean;
  }
  
  /**
   * Get partner discovery configuration.
   */
  async getPartnerDiscoveryConfig(): Promise<PartnerDiscoveryConfig> {
    const response = await this.client.get('/config/partner-discovery');
    this.checkResponse(response);
    return JSON.parse(response.body) as PartnerDiscoveryConfig;
  }
  
  /**
   * Set partner discovery configuration.
   */
  async setPartnerDiscoveryConfig(config: PartnerDiscoveryConfig): Promise<void> {
    const response = await this.client.put('/config/partner-discovery', config);
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Monitor Control
  // ─────────────────────────────────────────────────────────────
  
  /**
   * Restart the ProcessMonitor itself.
   */
  async restartMonitor(): Promise<void> {
    const response = await this.client.post('/monitor/restart');
    this.checkResponse(response);
  }
  
  // ─────────────────────────────────────────────────────────────
  // Private Helpers
  // ─────────────────────────────────────────────────────────────
  
  private checkResponse(response: { status: number; body: string; ok: boolean }): void {
    if (response.status === 403) {
      throw new PermissionDeniedException(response.body);
    }
    if (response.status === 404) {
      try {
        const body = JSON.parse(response.body);
        throw new ProcessNotFoundException(body.processId ?? 'unknown');
      } catch (e) {
        if (e instanceof ProcessNotFoundException) throw e;
        throw new ProcessNotFoundException('unknown');
      }
    }
    if (!response.ok) {
      throw new HttpException(response.status, response.body);
    }
  }
  
  private parseProcessStatus(data: Record<string, unknown>): ProcessStatus {
    return {
      id: data.id as string,
      name: data.name as string,
      state: (data.state as ProcessState) ?? 'unknown',
      enabled: data.enabled as boolean,
      autostart: data.autostart as boolean,
      isRemote: data.isRemote as boolean,
      pid: data.pid as number | undefined,
      lastStartedAt: data.lastStartedAt 
        ? new Date(data.lastStartedAt as string) 
        : undefined,
      lastStoppedAt: data.lastStoppedAt 
        ? new Date(data.lastStoppedAt as string) 
        : undefined,
      restartAttempts: (data.restartAttempts as number) ?? 0,
    };
  }
  
  private parseMonitorStatus(data: Record<string, unknown>): MonitorStatus {
    return {
      instanceId: data.instanceId as string,
      pid: data.pid as number,
      startedAt: new Date(data.startedAt as string),
      uptime: data.uptime as number,
      state: data.state as string,
      standaloneMode: (data.standaloneMode as boolean) ?? false,
      partnerInstanceId: data.partnerInstanceId as string | undefined,
      partnerStatus: data.partnerStatus as string | undefined,
      partnerPid: data.partnerPid as number | undefined,
      managedProcessCount: data.managedProcessCount as number,
      runningProcessCount: data.runningProcessCount as number,
    };
  }
}
